src/kernel/scheduler.zig

```diff
@@ -0,0 +1,583 @@
+
+        task.stack[STACK_SIZE - 9] = 0; // ebx
+        task.stack[STACK_SIZE - 8] = 0; // edx
+        task.stack[STACK_SIZE - 7] = 0; // ecx
+        task.stack[STACK_SIZE - 6] = 0; // eax
+
+        task.stack[STACK_SIZE - 5] = 0; // int_num
+        task.stack[STACK_SIZE - 4] = 0; // error_code
+
+        task.stack[STACK_SIZE - 3] = @ptrToInt(entry_point); // eip
+        task.stack[STACK_SIZE - 2] = gdt.KERNEL_CODE_OFFSET; // cs
+        task.stack[STACK_SIZE - 1] = 0x202; // eflags
+
+        // Point to top of the created stack above
+        task.esp = @ptrToInt(&task.stack[STACK_SIZE - 18]);
+
+        return task;
+    }
+
+    ///
+    /// Destroy the task. This will release the allocated PID and free the stack and self.
+    ///
+    /// Arguments:
+    ///     IN/OUT self: *Self - The pointer to self.
+    ///
+    pub fn destroy(self: *Self) void {
+        freePid(self.pid);
+        alloc.free(self.stack);
+        alloc.destroy(self);
+    }
+
+    ///
+    /// Save a new stack pointer for this task.
+    ///
+    /// Arguments:
+    ///     IN/OUT self: *Self - The pointer to self.
+    ///     IN esp: u32        - The new stack pointer to be saved.
+    ///
+    pub fn saveEsp(self: *Self, esp: u32) void {
+        self.esp = esp;
+    }
+};
+
+///
+/// Allocate a process identifier. If out of PIDs, then will panic. Is this occurs, will need to
+/// increase the bitmap.
+///
+/// Return: u32
+///     A new PID.
+///
+fn allocatePid() u32 {
+    return all_pids.setFirstFree() orelse panic(@errorReturnTrace(), "Out of PIDs", .{});
+}
+
+///
+/// Free an allocated PID. One must be allocated to be freed. If one wasn't allocated will panic.
+///
+/// Arguments:
+///     IN pid: u32 - The PID to free.
+///
+fn freePid(pid: u32) void {
+    all_pids.clearEntry(pid) catch panic(@errorReturnTrace(), "PID {} not allocated", .{pid});
+}
+
+///
+/// Round robin. This will pick the next task to be run from the queue. It will add the current
+/// task to the end of the queue and pop the next task from the front as set this as the current
+/// task. Interrupts are assumed disabled.
+///
+fn pickNextTask() void {
+    // Get the next task
+    if (tasks.pop()) |new_task_node| {
+        var next_task = new_task_node.data;
+        tasks.destroyNode(new_task_node, alloc);
+
+        // Now that we have the new task and deleted its node, there "will" be space for another
+        // so the allocation shouldn't fail as interrupts are disabled.
+        // So if there is a failure, then wow :P This will be the case for multiprocessor, but
+        // for a later time
+
+        // Put the current task back in the queue, at the back
+        var cur_elem = tasks.createNode(current_task, alloc) catch |e| {
+            panic(@errorReturnTrace(), "Error in pickNextTask, can't save previous task: {}\n", .{e});
+        };
+        tasks.prepend(cur_elem);
+
+        // Set the next task
+        current_task = next_task;
```

Comment:
Making this return the next task rather than set the global variable would make it more generic and usable in contexts when you don't want to set it

Reply:
So will `pickNextTask` then return a Task and then in the schedule function: `current_task = pickNextTask();`

Reply:
That's my thought

Reply:
Ok, it make sense now :) I was thinking just thinking that if pickNext would return the task to switch to, then at some point it will set the current_task, so why not just set it in the pickNext to reduce that step. All the switchTask function will do will return the stack pointer of the task as the interrupt handler will swap the task